dev/py-tests/requirements-sql.txt

```diff
@@ -0,0 +1,11 @@
+# PySpark dependencies for SQL tests
+
+numpy==1.26.2
```

Comment:
> The requirements file format allows for specifying dependency versions using logical operators (for example chardet>=3.0.4) or specifying dependencies without any versions. In this case the pip install -r requirements.txt command will always try to install the latest available package version. To be sure that the cache will be used, please stick to a specific dependency version and update it manually if necessary.

specify versions according to the suggestion in https://github.com/actions/setup-python?tab=readme-ov-file#caching-packages-dependencies

actually, I think maybe we should always specify the versions

Reply:
> actually, I think maybe we should always specify the versions

I agree with this, and this is something I tried to do in the PR I linked to just above, but several committers were against it.

When I look at the [number of PRs related to pinning dev dependencies][1] over the past three years, I wonder if committers still feel the same way today.

Not pinning development dependencies creates constant breakages that can pop up whenever an upstream library releases a new version. When we pin dependencies, by contrast, we choose when to upgrade and deal with the potential breakage.

[1]: https://github.com/apache/spark/pulls?q=is%3Apr+infra+pin

Reply:
Yeah. I don't have a great solution on this. We could have CI dedicated dep files maybe ... because now we have too many dependencies in CI with too many matrix ... but not sure .. At least, now I am not super against this idea..

Reply:
Would you be open to my making another attempt at the approach in #27928? (@zhengruifeng can also take this on if they prefer, of course.)

Basically, we have two sets of development dependencies:
- `requirements.txt`: direct dependencies only that are as flexible as possible; this is what devs install on their laptops
- `requirements-pinned.txt`: pinned dependencies derived automatically from `requirements.txt` using [pip-tools][1]; this is used for CI

I know this adds a new tool that non-Python developers may not be familiar with (pip-tools), but it's extremely easy to use, has been around a long time, and is in use by many large projects, the most notable of which is [Warehouse][2], the project that backs PyPI.

[1]: https://github.com/jazzband/pip-tools
[2]: https://github.com/pypi/warehouse

Reply:
@nchammas I just notice the previous discussion https://github.com/apache/spark/pull/27928.

I personally prefer using `requirements.txt` files with pinned versions, one reason is that the dependency is actually cached in docker file, and I was confused about the version used in CI from time to time, e.g.
we used the cached `RUN python3.9 -m pip install numpy pyarrow ...` before, and when pyarrow 13 released at 2023-8-23, I didn't know this release broke PySpark before the cached image was refreshed (at 2023-9-13).

But I don't feel very strong about it and defer to @HyukjinKwon and @dongjoon-hyun on this.